File: sample/src/main/java/de/rheinfabrik/heimdalldroid/network/oauth2/rxjava/TraktTvAuthorizationCodeGrant.java
package de.rheinfabrik.heimdalldroid.network.oauth2.rxjava;

import android.net.Uri;

import de.rheinfabrik.heimdall2.rxjava.grants.OAuth2AuthorizationCodeGrant;
import io.reactivex.Observable;
import java.net.URL;

import de.rheinfabrik.heimdall2.accesstoken.OAuth2AccessToken;
import de.rheinfabrik.heimdalldroid.network.TraktTvApiFactory;
import de.rheinfabrik.heimdalldroid.network.models.AccessTokenRequestBody;

/**
 * TraktTv authorization code grant as described in http://docs.trakt.apiary.io/#reference/authentication-oauth.
 */
public class TraktTvAuthorizationCodeGrant extends OAuth2AuthorizationCodeGrant {

    // Properties

    public String clientSecret;

    // OAuth2AuthorizationCodeGrant

    @Override
    public URL buildAuthorizationUrl() {
        try {
            return new URL(
                    Uri.parse("https://trakt.tv/oauth/authorize")
                            .buildUpon()
                            .appendQueryParameter("client_id", getClientId())
                            .appendQueryParameter("redirect_uri", getRedirectUri())
                            .appendQueryParameter("response_type", OAuth2AuthorizationCodeGrant.RESPONSE_TYPE)
                            .build()
                            .toString()
            );
        } catch (Exception ignored) {
            return null;
        }
    }

    @Override
    public Observable<OAuth2AccessToken> exchangeTokenUsingCode(String code) {
        AccessTokenRequestBody body = new AccessTokenRequestBody(
                code, getClientId(), getRedirectUri(), clientSecret, GRANT_TYPE
        );
        return TraktTvApiFactory.newApiServiceRxJava().grantNewAccessToken(body);
    }
}

File: sample/src/main/java/de/rheinfabrik/heimdalldroid/network/oauth2/coroutines/TraktTvAuthorizationCodeGrant.kt
package de.rheinfabrik.heimdalldroid.network.oauth2.coroutines

import android.net.Uri
import de.rheinfabrik.heimdall2.accesstoken.OAuth2AccessToken
import de.rheinfabrik.heimdall2.coroutines.grants.OAuth2AuthorizationCodeGrant
import de.rheinfabrik.heimdalldroid.network.TraktTvApiFactory
import de.rheinfabrik.heimdalldroid.network.models.AccessTokenRequestBody
import java.net.URL

class TraktTvAuthorizationCodeGrant(
    val clientSecret: String
) : OAuth2AuthorizationCodeGrant() {

    // OAuth2AuthorizationCodeGrant
    override suspend fun buildAuthorizationUrl(): URL =
        URL(
            Uri.parse("https://trakt.tv/oauth/authorize")
                .buildUpon()
                .appendQueryParameter("client_id", clientId)
                .appendQueryParameter("redirect_uri", redirectUri)
                .appendQueryParameter("response_type", RESPONSE_TYPE)
                .build()
                .toString()
        )

    override suspend fun exchangeTokenUsingCode(code: String): OAuth2AccessToken =
        TraktTvApiFactory.newApiServiceCoroutines().grantNewAccessToken(
            body = AccessTokenRequestBody(
                code, clientId, redirectUri, clientSecret, GRANT_TYPE
            )
        )
}
